server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java

```diff
@@ -0,0 +1,383 @@
+                    removePeer();
+                }
+            });
+        }
+
+        private void removePeer() {
+            final Peer removed = peersByAddress.remove(transportAddress);
+            assert removed == Peer.this;
+        }
+
+        private void requestPeers() {
+            assert holdsLock() : "PeerFinder mutex not held";
+            assert peersRequestInFlight == false : "PeersRequest already in flight";
+            assert active;
+
+            final DiscoveryNode discoveryNode = getDiscoveryNode();
+            assert discoveryNode != null : "cannot request peers without first connecting";
+
+            logger.trace("{} requesting peers from {}", this, discoveryNode);
+            peersRequestInFlight = true;
+
+            List<DiscoveryNode> knownNodes = getKnownPeers();
+
+            transportService.sendRequest(discoveryNode, REQUEST_PEERS_ACTION_NAME,
+                new PeersRequest(getLocalNode(), knownNodes),
+                new TransportResponseHandler<PeersResponse>() {
+
+                    @Override
+                    public PeersResponse read(StreamInput in) throws IOException {
+                        return new PeersResponse(in);
+                    }
+
+                    @Override
+                    public void handleResponse(PeersResponse response) {
+                        logger.trace("{} received {} from {}", Peer.this, response, discoveryNode);
+                        synchronized (mutex) {
+                            if (active == false) {
+                                return;
+                            }
+
+                            peersRequestInFlight = false;
+
+                            if (response.getMasterNode().isPresent()) {
+                                final DiscoveryNode masterNode = response.getMasterNode().get();
+                                if (masterNode.equals(discoveryNode) == false) {
+                                    startProbe(masterNode.getAddress());
+                                }
+                            } else {
+                                response.getKnownPeers().stream().map(DiscoveryNode::getAddress)
+                                    .forEach(PeerFinder.this::startProbe);
+                            }
+                        }
+
+                        if (response.getMasterNode().equals(Optional.of(discoveryNode))) {
+                            // Must not hold lock here to avoid deadlock
+                            assert holdsLock() == false : "PeerFinder mutex is held in error";
+                            onActiveMasterFound(discoveryNode, response.getTerm());
+                        }
+                    }
+
+                    @Override
+                    public void handleException(TransportException exp) {
+                        peersRequestInFlight = false;
+                        logger.debug("PeersRequest failed", exp);
```

Comment:
add Peer.this to message?

Reply:
Sure, see e5414a170acd19535823c02537f6e68dd702b7a1